nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java

```diff
@@ -0,0 +1,293 @@
+                    flowFile = session.write(flowFile, out -> {
+                        out.write(bodyBytes);
+                    });
+                    logger.debug("Elasticsearch document " + docId + " fetched, routing to success");
+
+                    // emit provenance event
+                    final long millis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                    session.getProvenanceReporter().fetch(flowFile, url.toExternalForm(), millis);
+                    session.transfer(flowFile, REL_SUCCESS);
+                } else {
+                    logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                            new Object[]{index, docType, docId});
+
+                    // We couldn't find the document, so penalize it and send it to "not found"
+                    // If there was an incoming flow file, penalize and transfer. Otherwise just remove it
+                    if (context.hasIncomingConnection()) {
+                        flowFile = session.penalize(flowFile);
+                        session.transfer(flowFile, REL_NOT_FOUND);
+                    } else {
+                        session.remove(flowFile);
+                    }
+                }
+
+            } else {
+                if (statusCode == 404) {
+                    logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                            new Object[]{index, docType, docId});
+
+                    // We couldn't find the document, so penalize it and send it to "not found"
+                    // If there was an incoming flow file, penalize and transfer. Otherwise just remove it
+                    if (context.hasIncomingConnection()) {
+                        flowFile = session.penalize(flowFile);
+                        session.transfer(flowFile, REL_NOT_FOUND);
+                    } else {
+                        session.remove(flowFile);
+                    }
+                } else if (context.hasIncomingConnection()) {
+                    // 5xx -> RETRY
+                    if (statusCode / 100 == 5) {
+                        logger.info("Elasticsearch returned code {} with message {}, transferring flow file to retry", new Object[]{statusCode, getResponse.message()});
+                        flowFile = session.penalize(flowFile);
+                        session.transfer(flowFile, REL_RETRY);
+                    } else {
+                        // 1xx, 3xx, 4xx -> NO RETRY
+                        logger.info("Elasticsearch returned code {} with message {}, transferring flow file to failure", new Object[]{statusCode, getResponse.message()});
+                        session.transfer(flowFile, REL_FAILURE);
+                    }
+                } else {
+                    logger.info("Elasticsearch returned code {} with message {}", new Object[]{statusCode, getResponse.message()});
```

Comment:
There should be more robust handling here. For instance, where the processor has no incoming connections and hits a 4xx failure, should the processor yield? In general, this means the server is failing and we should probably not hammer it with requests. Also there should be a warning/error saying that the server is down.